Fix all outstanding lints and document expectations (#2457)
rust-lang/rustup: CONTRIBUTING.md

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Test it: `cargo test`

> 4. Lint it: `cargo +nightly clippy --all-targets -- -D warnings`

Reviewer: How come we use nightly to lint but not nightly to test?

Author: nightly often has better lints, or lints that tell us about things that are about to become an issue. In GitHub Actions we test with beta, which is a bit of a tradeoff between stability and features. While for releases we release with stable. For contributors generally they can build with whatever they want, but I'd prefer the benefit of checking with the latest lints, and thats very easy to do.

5. Commit your changes: `git commit -am 'Add some feature'`
6. Push to the branch: `git push origin my-new-feature`

> 7. Submit a pull request :D

Reviewer: We could advocate for github cli. :P

`gh pr create`

Author: Please do suggest a PR for that, I haven't looked into accessibility for it across platforms etc.

`std::env::current_dir` and `std::process::exit` permitting threaded tests of
the CLI logic; use `process()` rather than those APIs directly.

### Clippy lints

> We don't enforce lint-free-ness in the gate, because it is a moving target, but

Reviewer 2: "out the gate"?

we do ask that people run `cargo +nightly clippy --all-targets -- -D warnings`

> before submitting their code, to help keep things generally tidy.

Reviewer 2: Overall para feels slightly ambiguous. What does "please run it but we don't enforce it" really mean?

Note that ideally lint checking is done on at least Linux and Windows, as
OS-conditional code is a common source of unused imports and other small lints,

> which can build up over time.

Reviewer 2: Is this directed at a contributor? I am reading this as "you should check linting on both Linux and Windows" but I am not sure how I would actually do that without a Windows machine. Do you mean examining the CI output?

`std::env::current_dir` and `std::process::exit` permitting threaded tests of
the CLI logic; use `process()` rather than those APIs directly.

### Clippy lints

> We don't enforce lint status in the checks done by GitHub Actions, because

Reviewer: ```suggestion
We do not enforce lint status in the checks done by GitHub Actions, because
```
`don't` is more useful on normal conversation, not talking.

> clippy is a moving target that can make it hard to merge for little benfit.

Reviewer 2: ```suggestion
clippy is a moving target that can make it hard to merge for little benefit.
```

Reviewer: ```suggestion
clippy is a moving target which makes it hard to merge for little benefit.
```

We do ask that contributors keep the clippy status clean themselves.

Minimally, run `cargo +nightly clippy --all-targets -- -D warnings` before
submitting code.

Regular contributors or contributors to particularly OS-specific code should
also make sure that their clippy checking is done on at least Linux and Windows,
as OS-conditional code is a common source of unused imports and other small
lints, which can build up over time.

For developers using BSD/Linux/Mac OS, there are Windows VM's suitable for such
development tasks for use with virtualbox and other hypervisors are downloadable
from
[Microsoft](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/).
Similarly, there are many Linux and Unix operating systems images available for
developers whose usual operating system is Windows. Currently Rustup has no Mac
OS specific code, so there should be no need to worry about Mac VM images.

> Clippy is also run in GitHub Actions, in the 'General Checks / Checks` build

Reviewer 3: ```diff
-Clippy is also run in GitHub Actions, in the 'General Checks / Checks` build
+Clippy is also run in GitHub Actions, in the `General Checks / Checks` build
```

Reviewer: This is already merged, I think you can just send a pull request for this.

Reviewer 3: Feel free to do it on my behalf.
